### ISSUE TITLE
Some unsafe characters are not URL encoded when they are provided as query params

### ISSUE THREAD

**Issue by danbars** [a85/POSTMan-Chrome-Extension#13380]
### Is there an existing issue for this?

- [X] I have searched the tracker for existing similar issues and I know that duplicates will be closed

### Describe the Issue

The following characters must be URL-encoded when they appear as query-param values:
- Ampersand (&)
- Equals sign (=)
- Plus sign (+)
- Question mark (?)
- Hash/Number sign (#)

The following characters are considered unsafe and should be encoded as well, even though they are not reserved:
- Space (encoded as %20 or +)
- Less than (<) and greater than (>)
- Double quotes (")
- Percent sign (%)

In Postman, when a query param value that uses any of the above characters is provided, it is automatically being URL-encoded. The only 2 exceptions are the percent `%` and plus `+` characters

The js function [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#description) describes how caracters should be encoded when used in URL values:  
> It escapes all characters except: `A–Z a–z 0–9 - _ . ! ~ * ' ( )`

[RFC-3986, section 2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2) specifies the reserved characters, and shows that the plus sign has to be encoded. [Section 2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-2.4) states that:  
> Because the percent ("%") character serves as the indicator for
   percent-encoded octets, it must be percent-encoded as "%25" for that
   octet to be used as data within a URI.

### Steps To Reproduce

1. Create a GET request
2. Go to Params tab
3. Enter 4 params:
   1. key=`space`, value=`x y`
   2. key=`ampersand`, value=`x&y`
   3. key=`percent`, value=`x%y`
   4. key=`plus`, value=`x+y`
5. Open the code snippet tab on the right, and observe the cURL command: `curl --location 'https://foo.com?space=x%20y&ampersand=x%26y&percent=x%y&plus=x+y'`
   1. The space and ampersand characters were URL-encoded
   2. The percent and plus characters were not URL-encoded

Note: all other characters mentioned above are also being encoded. They are not mentioned in the reproduction steps here just for simplicity.

### Screenshots or Videos

![image](https://github.com/user-attachments/assets/f4f2d3cc-12e1-408b-8e13-7c5213cefdd9)

### Operating System

macOS

### Postman Version

11.23.3

### Postman Platform

Postman App

### User Account Type

Signed In User

### Additional Context?

There is [issue 8626](https://github.com/postmanlabs/postman-app-support/issues/8626) that describes part of the issue mentioned here, but it doesn't cover the percentage character.
Moreover, it is over 4 years old, and it seems that because of misclassification it no longer gets any attention from Postman team.
The confusion in that other ticket stems from the fact that plus sign is both:
1. An unsafe character that has to be encoded into `%2B`.
2. The result of encoding space character which may be encoded into `%20` or `+`.

**Comment by jonathanhaviv** [comment]
Hi @danbars thank you for bringing this to our attention and for the detailed report. We will take a look at this and let you know once we have released a fix for this.

**Comment by ASk1** [comment]
and semicolon (%3B) and comma (%2C) as well - they encoded in curl snippet but sent as is by postman itself

**Comment by danbars** [comment]
@ASk1 , good catch. I was counting on the cURL representation and didn't check how it's actually being sent.

**Comment by SvendK** [comment]
Also equals sign ('=') in parameter values are not encoded. It is in the cURL representation, but in the console the = is visible.
And like OP writes, my % signs (wildcard character in the DepartmentCode LIKE parts) are not converted to %25

So besides the encoding, could you also add to the 'todo' list, to make sure the cURL representation have the actual output? I'd guess this would be the same full URL as what is shown in Console after a request? I might be wrong :-)

Maybe each parameter should have a checkbox so the user can say "this value should (not) be urlencoded / this value is already urlencoded"? And then make sure Postman encodes all unsafe characters, if it's supposed to? :)

**My "Filter" query parameter's value**

> transfered = 0 AND Date >= '2024-10-01T00:00:00' AND Date <= '2024-10-31T23:59:59' AND (DepartmentCode LIKE '%70%' OR DepartmentCode LIKE '%30%' OR DepartmentCode LIKE '%90%' OR DepartmentCode LIKE '%80%' OR DepartmentCode LIKE '%100%' OR DepartmentCode LIKE '%56%')`

**Code snippet cURL representation**

> curl --location 'https:// (...) /timeledgerentry?limit=3&offset=0&OrderFieldName=Date&OrderDirection=Desc&Filter=transfered%20%3D%200%20AND%20Date%20%3E%3D%20%272024-10-01T00%3A00%3A00%27%20AND%20Date%20%3C%3D%20%272024-10-31T23%3A59%3A59%27%20AND%20(DepartmentCode%20LIKE%20%27%70%%27%20OR%20DepartmentCode%20LIKE%20%27%30%%27%20OR%20DepartmentCode%20LIKE%20%27%90%%27%20OR%20DepartmentCode%20LIKE%20%27%80%%27%20OR%20DepartmentCode%20LIKE%20%27%100%%27%20OR%20DepartmentCode%20LIKE%20%27%56%%27)' \
(... headers omitted)

**Request in Console**

> GET https:// (...) /timeledgerentry?limit=3&offset=0&OrderFieldName=Date&OrderDirection=Desc&Filter=transfered%20=%200%20AND%20Date%20%3E=%20%272024-10-01T00:00:00%27%20AND%20Date%20%3C=%20%272024-10-31T23:59:59%27%20AND%20(DepartmentCode%20LIKE%20%27%70%%27%20OR%20DepartmentCode%20LIKE%20%27%30%%27%20OR%20DepartmentCode%20LIKE%20%27%90%%27%20OR%20DepartmentCode%20LIKE%20%27%80%%27%20OR%20DepartmentCode%20LIKE%20%27%100%%27%20OR%20DepartmentCode%20LIKE%20%27%56%%27)


If I urlencode my original Filter value and insert that as the Postman Params "Filter" value instead, then the Params Filter value, the cURL Filter parameter value and the Console Filter value are the same - thanks to the missing % encoding. So to send my filter string to the API via Postman, I need to urlencode it myself, which I believe is not right? And if you fix % encoding and not = encoding, then I need to encode only the equal signs and not the whole string.

**The urlencoded Filter parameter/query parameter**

> Filter=transfered%20%3D%200%20AND%20Date%20%3E%3D%20%272024-10-01T00%3A00%3A00%27%20AND%20Date%20%3C%3D%20%272024-10-31T23%3A59%3A59%27%20AND%20%28DepartmentCode%20LIKE%20%27%2570%25%27%20OR%20DepartmentCode%20LIKE%20%27%2530%25%27%20OR%20DepartmentCode%20LIKE%20%27%2590%25%27%20OR%20DepartmentCode%20LIKE%20%27%2580%25%27%20OR%20DepartmentCode%20LIKE%20%27%25100%25%27%20OR%20DepartmentCode%20LIKE%20%27%2556%25%27%29


I'm looking forward to any progress on this :-)

Best regards,
SvendK

**Comment by danbars** [comment]
@jonathanhaviv any update on this issue?

**Comment by Skorpyon** [comment]
Especially it annoying when sending timestamps!

<img width="430" height="115" alt="Image" src="https://github.com/user-attachments/assets/af00d4b3-4baf-488b-9cc1-8e156381503b" />

It sent directly with non-encoded "+" char and it interpreted as space in web server:

<img width="480" height="68" alt="Image" src="https://github.com/user-attachments/assets/fa0b6424-88ad-4432-b95a-0a77bc6603a0" />

That is not valid ISO format and backend validation rejecting timestaps.
